src/App.js

```diff
@@ -1,26 +1,99 @@
+import React, {useEffect, useRef, useState} from 'react';
+import {ACCESS_KEY, API_URL} from './credentials';
+import axios from 'axios';
 import './App.css';
+import Pagination from "./components/pagination";
 
 function App() {
-  return (
-    <div className="App">
-      <header className="App-header">
-        <img src={logo} className="App-logo" alt="logo" />
-        <p>
-          Edit <code>src/App.js</code> and save to reload.
-        </p>
-        <a
-          className="App-link"
-          href="https://reactjs.org"
-          target="_blank"
-          rel="noopener noreferrer"
-        >
-          Learn React
-        </a>
-      </header>
-    </div>
-  );
+    const [images, setImages] = useState([]);
+    const [page, setPage] = useState(1);
+    const loader = useRef(null);
+    const [query, setQuery] = useState('')
+
+
+    const fetchImages = () => {
+        axios.get(`${API_URL}/photos/?client_id=${ACCESS_KEY}&per_page=10&page=${page}`)
+            .then(res => setImages([...images, ...res.data]));
+    };
+
+    const searchImages = () => {
+        axios.get(`${API_URL}/search/photos/?client_id=${ACCESS_KEY}&per_page=10&page=${page}&query=${query}`)
+            .then(res => setImages([...images, ...res.data.results]));
+    };
+
+    const handleObserver = entities => {
+         const target = entities[0];
+
+        /*if (target.intersectionRatio > 0) {
+            setPage(page => page + 1)
+        }*/
+    };
+
+    useEffect(() => {
+        const options = {
+            root: null,
+            rootMargin: '0px',
+            threshold: 1.0,
+        }
+        const observer = new IntersectionObserver(handleObserver, options);
+        if (loader.current) {
+            observer.observe(loader.current);
+        }
+    }, [])
+
+    useEffect(() => {
+        if (query) {
+            searchImages();
+        } else {
+            fetchImages();
+        }
+    }, [page, query]);
+
+    const handleClick =  (newQuery) => {
+        if(newQuery !== query){
+            setImages([]);
+            setPage(1);
+        }
+        setQuery(newQuery);
+        setQuery(newQuery);
+    }
+
+    const totalPages = 40;
+
+   const changeCurrentPage = (pageNumber) => {
+        console.log("click page #" + pageNumber)
+       setImages([]);
+       setPage(pageNumber)
+    }
+
+    return (
+        <div className="container">
+            <header className="header">
+                <h1>Fancy Gallery</h1>
+            </header>
+            <div className="tags">
+                <button onClick={()=>handleClick('cats')}>Cats</button>
+                <button onClick={()=>handleClick('dogs')}>Dogs</button>
+                <button onClick={()=>handleClick('react')}>React</button>
+                <button onClick={()=>handleClick('')}>Random</button>
+            </div>
+            <Pagination
+                totalPages={totalPages}
+                changeCurrentPage={changeCurrentPage}
```

Comment:
you can use here setPage, if you remove console.log and setImages from changeCurrentPage handler